### ISSUE TITLE
How to allow character '&' without convert to '&amp;'

### ISSUE THREAD

**Issue by rifkiharahap** [jhy/jsoup#372]
i have data from http post data like this :

```
"billingCode=820131114000031&custName=Josh & Mark&numAccount=25800&parameterId=2&traceId=820131114000031"
```

if i use jsoup :

``` Java
String postDataori = Jsoup.clean("billingCode=820131114000031&custName=Josh & Mark&numAccount=25800&parameterId=2&traceId=820131114000031", Whitelist.basic());
System.out.println("PostData  : " + postDataori);
```

and the output like this :

``` Java
PostData  : billingCode=820131114000031&amp;custName=Josh &amp; Mark&amp;numAccount=25800&amp;parameterId=2&amp;traceId=820131114000031
```

my question is how to allow character '&' without convert to `&amp;` ?

**Comment by isapir** [comment]
I'm sorry but this seems like a bug to me, or at least a missing feature, rather than a simple question.  I have read comments on StackOverflow as well as the mailing list.  IMO this ticket should be re-opened.

In most cases in URLs the Ampersand symbol is used as a delimiter and not an entity.

This is what I need:
<a href="https://www.google.com/search?site=&source=hp&q=jsoup">Search google for jsoup</a>

This is what JSoup produces:
<a href="https://www.google.com/search?site=&amp;source=hp&amp;q=jsoup">Broken link</a>

As for the example with "&copy=" that I've seen on the mailing list, that would not be parsed as an entity as it's missing the semi-colon at the end.  The entity would be "&copy;"

Any thoughts?

Thanks!

**Comment by isapir** [comment]
It seems to me that the fix should be at 
https://github.com/jhy/jsoup/blob/master/src/main/java/org/jsoup/nodes/Entities.java#L188
where 

    accum.append("&amp;");

should be changed to

    if (!inAttribute)
        accum.append("&amp;");

**Comment by cketti** [comment]
The HTML5 spec allows for some leniency when parsing which is why not encoding `&` as `&amp;` leads to the desired behavior most of the time. But always encoding ampersands is certainly not wrong.

See https://www.w3.org/TR/html5/syntax.html#attributes-0 and https://www.w3.org/TR/html5/syntax.html#character-reference-in-attribute-value-state

**Comment by isapir** [comment]
@cketti My issue is not with Parsing; It's with rendering of the HTML when calling Document's toString() or html().

As I wrote above, I need to render links that work, so the delimiter of the Query String is a single `&` symbol, and not an HTML entity, e.g.
https://www.google.com/search?source=hp&q=jsoup

But JSoup always produces the delimiter as HTML entities, which leads to broken links, e.g.
https://www.google.com/search?source=hp&amp;q=jsoup

There has to be a way to render the HTML with single `&`.  This issue has affected many others judging by the mailing list and StackOverflow, and they resorted to doing a (pseudo-code) `Replace(html, "&amp;", "&")`

The best solution IMO is to add another `EscapeMode` that will be more permissive, and will not enforce entities where they not desired.

**Comment by cketti** [comment]
If you output HTML the values should be encoded. If you need the decoded attribute value, extract it using e.g. `Node#attr(String)`.

```java
Document document = Jsoup.parse("<a href=\"https://www.google.com/search?source=hp&amp;q=jsoup\">link</a>");
Element aElement = document.select("a").first();
assertEquals("https://www.google.com/search?source=hp&q=jsoup", aElement.attr("href"));
```

**Comment by isapir** [comment]
It certainly should not be encoded in all cases.  I am not outputting it for display in a browser.  I am using JSoup to manipulate HTML code and send it via email:

1. Read original HTML
2. Modify links to add tracking code to them
3. Render modified HTML and send via email

But when the email is sent then all of my links are broken because all of the URL delimiters are encoded.

**Comment by isapir** [comment]
@jhy I can provide a patch if you are open to it.  

IMO the best solution would be to add a new EscapeMode that would either be more permissive, or do nothing at all.  That will not change anything for current users, but will allow users like me to produce HTML with links that do not break the Query String.

Thank you for your consideration.

**Comment by mjclemente** [comment]
@jhy Just noting that I strongly agree with @isapir on this. 

Jsoup is **amazing**, but the automatic encoding of ampersands in links and attributes, without the ability to bypass this is very frustrating. We, too, resort to the rather hacky: `doc.toString().replace("&amp;", "&");`.

I too would like to be able to `parse()` and then output `toString()` without all ampersands in attributes being encoded.

Please considering reopening this and allowing modification of this. Thank you!

**Comment by dorward** [comment]
> As I wrote above, I need to render links that work, 

Then `&amp;` should not be a problem. 

Any HTML parser will parse the `<a>` element, with the `href` attribute, hit `&amp;`, understand that `&amp;` means "This is the character reference for an ampersand" and then insert a `&` into the value of the DOM attribute node.

The only time you would have a problem is if you tried to use the HTML source code as a plain text URL … which it isn't.

**Comment by mjclemente** [comment]
@dorward The automatic encoding can cause other issues.

Our use case for Jsoup is to parse code for specific tags (in this case, scripts) and extract them, so they can be stored separately. After parsing the code with Jsoup, and extracting the tags, I attempted to set up a string length comparison (original input vs sum of extracted code), in order to validate the user input and make sure no information was unintentionally left out in parsing.

While Jsoup makes it amazingly easy to extract the tags, by altering them (via the ampersands), I am unable to compare the original code length with the extracted code length (unless I run the hacky replace - `toString().replace("&amp;", "&");`).

Just another use case for the possibility of bypassing the automatic encoding, if possible.

**Comment by mjclemente** [comment]
Also wanted to note that, it looked like the unencoded ampersand in attributes, in most cases, is valid according to the HTML5 spec: https://www.w3.org/TR/html5/syntax.html#tokenizing-character-references 

And here's another helpful breakdown: https://mathiasbynens.be/notes/ambiguous-ampersands

**Comment by isapir** [comment]
> Then `&amp;` _should_ not be a problem.

@dorward That's the problem here -- everyone keeps telling us that it _should_ work, but anyone who has ever tested it or tried to make it work knows that it doesn't.

Any link generated by JSoup that has more than one URL parameter is broken.  Ampersands should not be encoded in attributes.

**Comment by cketti** [comment]
@mjclemente: You can't just compare (decoded) content with encoded data and expect them to be identical. Your string length check is most likely flawed, not the way Jsoup works.
I'm happy to comment on your approach if you post a question with your current code to Stack Overflow.

@isapir: Whatever consumes the HTML output from Jsoup *must* decode character references (e.g. `&amp;` to `&`) when parsing. It's part of the HTML specification. If that doesn't happen, then that application is broken.
Now, you could work around that bug on the receiving side by making Jsoup not use character references unless strictly necessary. But the rules that allow unencoded ampersands have most likely been added because there's so much crappy HTML around that still needs to be supported, not because it's a good idea. That rule makes parsing harder and lets broken HTML generators and parsers get away with being broken for far too long. In most cases specifications contain text along the lines of "you still need to support this weirdness when reading, but you must encode these characters when generating a document". It's a bit unfortunate that the HTML specification does not.
I for one believe Jsoup is doing the right thing by always encoding `&`. It'll make the world of HTML a better place in the long run.

Jsoup's DOM API should be powerful enough for you to write your own HTML serializer if you really want to go that route. There's no need to change Jsoup for that.

**Comment by isapir** [comment]
> It's part of the HTML specification. If that doesn't happen, then that application is broken.

**Comment by mjclemente** [comment]
@cketti thank you. I've got the length check working fine, it's just, like I said, that `toString().replace("&amp;", "&");)` feels hacky. The only issue was it was Jsoup's _unnecessary_ encoding of ampersands when outputting a document that it has parsed. If the ampersand was not encoded in the original, and it conforms to the HTML5 spec, then there should at least be an option for not encoding it in the output. 

With regard to: 

> It's part of the HTML specification. If that doesn't happen, then that application is broken.

That's not true, as I pointed out:

> Also wanted to note that, it looked like the unencoded ampersand in attributes, in most cases, is valid according to the HTML5 spec: https://www.w3.org/TR/html5/syntax.html#tokenizing-character-references

Simply put, **unambiguous ampersands are valid**. Changing them is unnecessary.

**Comment by isapir** [comment]
> `toString().replace("&amp;", "&")` feels hacky

@mjclemente That's because it is indeed very much hacky

**Comment by mjclemente** [comment]
Hahaha. True. I fully acknowledge that. Which is why I was hoping Jsoup would have an option to allow valid ampersands to remain unencoded. But if that's not gonna happen, at least my hacky code is getting the job done 👍

**Comment by cketti** [comment]
@isapir: Maybe the links generated by you break because of something else. Using `&amp;` in `href` attributes works just fine with all email clients I'm using. Which is not too surprising since all of them use HTML engines that are also used by major browsers.

Feel free to test this yourself:
```
From: <alice@example.org>
To: <bob@example.org>
Subject: Ampersand in HTML attribute encoded as &amp;
Date: Fri, 24 Feb 2017 03:14:00 +0100
MIME-Version: 1.0
Content-Type: text/html; charset=utf-8

<html>
<head></head>
<body>
  This Link should work just fine:
  <a href="https://www.google.com/search?source=hp&amp;q=jsoup">Link</a>
</body>
</html>
```

**Comment by matthew-heyner** [comment]
Just ran into this problem myself. Not sure why we can't have some level of customizability here to allow consumers of jsoup to configure it to tailor our needs.
